src/Extensions.Hosting.AsyncInitialization/Hosting/AsyncInitializationHostExtensions.cs

```diff
@@ -67,18 +47,60 @@ public static async Task InitAndRunAsync(this IHost host, CancellationToken canc
         {
             if (host == null) throw new ArgumentNullException(nameof(host));
 
-            await using (host as IAsyncDisposable)
+            await using var scope = host.Services.CreateAsyncScope();
+            var rootInitializer = scope.ServiceProvider.GetService<RootInitializer>()
+                ?? throw new InvalidOperationException("The async initialization service isn't registered, register it by calling AddAsyncInitialization() on the service collection or by adding an async initializer.");
+
+
+            await using (host.AsAsyncDisposable().ConfigureAwait(false))
             {
                 try
                 {
-                    await host.InitAsync(cancellationToken).ConfigureAwait(false);
+                    await rootInitializer.InitializeAsync(cancellationToken).ConfigureAwait(false);
                     await host.StartAsync(cancellationToken).ConfigureAwait(false);
                     await host.WaitForShutdownAsync(cancellationToken).ConfigureAwait(false);
                 }
                 finally
                 {
-                    await host.TeardownAsync(CancellationToken.None).ConfigureAwait(false);
+                    await rootInitializer.TeardownAsync(CancellationToken.None).ConfigureAwait(false);
+                }
+            }
+        }
+
+        // wraps an IHost instance in an IAsyncDisposable 
+        private static IAsyncDisposable AsAsyncDisposable(this IHost host)
+        {
+            if (host == null) throw new ArgumentNullException(nameof(host));
+            return AsyncDisposableHostWrapper.Create(host);
+        }
+
+        // IAsyncDisposable wrapper for IHost
+        private class AsyncDisposableHostWrapper : IHost, IAsyncDisposable
+        {
+            private readonly IHost _host;
+            public AsyncDisposableHostWrapper(IHost host)
+            {
+                _host = host ?? throw new ArgumentNullException(nameof(host));
+            }
+
+            public static AsyncDisposableHostWrapper Create(IHost host) => new(host);
```

Comment:
What's the point of this? Why not use the constructor directly?

Reply:
Using a factory would have allowed for creating a singleton. Decided against it but left the Create() method. But you're right, it's not needed.